torch_xla/csrc/torch_util.h

```diff
@@ -16,24 +16,27 @@ namespace torch_xla {
 // Unpack SymInt objects into their building blocks
 struct SymIntElements {
  public:
-  SymIntElements(c10::SymInt& size) { SetSymIntNodeElements(size); }
+  SymIntElements(c10::SymInt& size) { AddSymIntNodeElements(size); }
   SymIntElements(c10::SymIntArrayRef& size) {
     std::vector<c10::SymInt> _sizes = torch::lazy::ToVector<c10::SymInt>(size);
     for (auto& _size : _sizes) {
-      SetSymIntNodeElements(_size);
+      AddSymIntNodeElements(_size);
     }
   }
-  std::vector<torch::lazy::NodePtr> GetNodes() { return size_nodes_; }
-  std::vector<int64_t> GetUpperBounds() { return upper_bounds_; }
-  std::vector<bool> GetDynamicDims() { return dynamic_dims_; }
+  std::unordered_map<size_t, torch::lazy::NodePtr> GetNodeMap() const {
+    return size_node_map_;
+  }
+  std::vector<int64_t> GetUpperBounds() const { return upper_bounds_; }
+  std::vector<bool> GetDynamicDims() const { return dynamic_dims_; }
+  torch::lazy::NodePtr GetNode(size_t index) const;
   void SetUpperBound(int64_t index, int64_t upper_bound) {
     XLA_CHECK_GT(upper_bounds_.size(), index);
     upper_bounds_[index] = upper_bound;
   }
 
  private:
-  void SetSymIntNodeElements(c10::SymInt& size);
-  std::vector<torch::lazy::NodePtr> size_nodes_;
+  void AddSymIntNodeElements(c10::SymInt& size);
+  std::unordered_map<size_t, torch::lazy::NodePtr> size_node_map_;
```

Comment:
I think we **need** a comment here. I believe @miladm 's intention here was to only keep track of dimension indices that map to size nodes.

Reply:
we could probably get away with using a dense vector with nullptr gaps since we always know a rank ahead of time to allocate it. it would probably make `SymIntElements::GetNode` simpler, but no strong preferences.